src/core/thread/mesh_forwarder.cpp

```diff
@@ -2016,6 +2016,11 @@ void MeshForwarder::HandleDataRequest(const Mac::Address &aMacSource, const Thre
     child->SetLastHeard(Timer::GetNow());
     child->ResetLinkFailures();
 
+    if (aMacSource.mLength == sizeof(aMacSource.mShortAddress))
+    {
+        mSourceMatchController.SetSrcMatchAsShort(*child, true);
```

Comment:
You are correct that we may have one wasted data poll but it will happen only in the rare scenario from the description (i.e., child gets the "Child ID Response" and parent misses the ack to it). I think that may be fine.

The check above is in addition to the existing code where we switch to using short address after a successful indirect tx to the child. So in the common case where the Child ID Resp. exchange is successful, we won't have any wasted data request.  

I am not sure about switching to short address after first (successful/failed) indirect tx. If we do this, we are basically limiting the  "Child ID response" to a single tx attempt. If the first attempt is unsuccessful, parent switches to using short address while child is still on long address. All later tx attempts fail and child needs to start with a new "Child ID Request".